mount/registry_test.go

```diff
@@ -0,0 +1,131 @@
+package mount
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+var _ Mount = (*MockMount)(nil)
+
+type MockMount struct {
+	Val      string
+	URL      *url.URL
+	StatSize uint64
+}
+
+func (m *MockMount) Fetch(_ context.Context) (io.ReadCloser, error) {
+	r := io.NopCloser(strings.NewReader(m.Val))
+	return r, nil
+}
+
+func (m *MockMount) Info() Info {
+	return Info{
+		Kind: MountKindRemote,
+		URL:  m.URL,
+	}
+}
+
+func (m *MockMount) Stat() (Stat, error) {
+	return Stat{
+		Exists: true,
+		Size:   m.StatSize,
+	}, nil
+}
+
+type MockMountFactory1 struct{}
+
+func (mf *MockMountFactory1) Parse(u *url.URL) (Mount, error) {
+	vals, err := url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	statSize, err := strconv.ParseUint(vals["size"][0], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MockMount{
+		Val:      u.Host,
+		URL:      u,
+		StatSize: statSize,
+	}, nil
+}
+
+type MockMountFactory2 struct{}
+
+func (mf *MockMountFactory2) Parse(u *url.URL) (Mount, error) {
+	vals, err := url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	statSize, err := strconv.ParseUint(vals["size"][0], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MockMount{
+		Val:      u.Host,
+		URL:      u,
+		StatSize: statSize * 2,
+	}, nil
+}
+
+func TestRegistry(t *testing.T) {
+	m1StatSize := uint64(1234)
+	m2StatSize := uint64(5678)
+
+	// create a registry
+	r := Registry{
+		m: make(map[string]MountFactory),
+	}
+
+	// create & register mock mount factory 1
+	u := fmt.Sprintf("http://host1:123?size=%d", m1StatSize)
+	url, err := url.Parse(u)
+	require.NoError(t, err)
+	m1 := &MockMountFactory1{}
+	require.NoError(t, r.Register(url.Scheme, m1))
```

Comment:
Better to hardcode the expected fixed value: Register("http", ...). Otherwise the test becomes rather circular.